modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/components/A11y/A11yChecker.ts

```diff
@@ -0,0 +1,398 @@
+	mutations?: Record<MutationRecordType, Mutation>;
+
+	/**
+	 * Targets is a list or element that represents the subtree(s) to be
+	 * analyzed.
+	 */
+	targets: Array<string>;
+}
+
+export class A11yChecker {
+	private callback: A11yCheckerOptions['callback'];
+	private scheduler: Scheduler<Node>;
+	private observers: Array<{
+		target: string;
+		mutation: MutationObserver;
+	}>;
+	private mutations?: Record<MutationRecordType, Mutation>;
+	readonly axeOptions: RunOptions;
+	readonly denylist: A11yCheckerOptions['denylist'];
+
+	constructor({
+		axeOptions,
+		callback,
+		denylist,
+		mutations,
+		targets,
+	}: A11yCheckerOptions) {
+		const defaultOptions = {
+			absolutePaths: true,
+			reporter: 'v2',
+		} as const;
+
+		this.axeOptions = axeOptions ? axeOptions : defaultOptions;
+
+		this.callback = callback;
+		this.denylist = denylist;
+
+		// Scheduler can be a singleton to allow multiple instances of
+		// A11yChecker to share the same queue to avoid bottlenecks and
+		// performance issues.
+
+		this.scheduler = new Scheduler();
+
+		this.mutations = mutations;
+
+		this.observers = targets.map((target: string) => ({
+			mutation: new MutationObserver(this.mutationCallback.bind(this)),
+			target,
+		}));
+	}
+
+	private async run(target: HTMLElement) {
+
+		// Only perform the analysis if the element exists on the screen
+
+		if (!document.body.contains(target)) {
+			return;
+		}
+
+		const results = await axe.run(target, this.axeOptions);
+
+		this.callback(results);
+	}
+
+	private recordCallback(target: Node) {
+		const hasCallbackScheduled = this.scheduler.hasCallback(
+			(scheduled) => scheduled.target === target
+		);
+
+		if (!hasCallbackScheduled) {
+			this.scheduler.scheduleCallback(this.run.bind(this), target);
+		}
+	}
+
+	private mutationCallback(records: Array<MutationRecord>) {
+		records.forEach((record) => {
+			if (this.mutations) {
+				const condition = this.mutations[record.type];
+
+				if (condition && hasValidMutation(record, condition)) {
+					this.recordCallback(record.target);
+				}
+			}
+			else {
+				this.recordCallback(record.target);
+			}
+		});
+	}
+
+	observe() {
+		this.observers.forEach(({mutation, target}) => {
+			const element = document.querySelector(target);
+
+			if (element) {
+				mutation.observe(element, {
+					attributes: true,
+					childList: true,
+					subtree: true,
+				});
+			}
+			else {
+				console.error(
+					`A11y: Target ${target} was not found in the DOM.`
+				);
+			}
+		});
+	}
+
+	unobserve() {
+		this.scheduler.cancelHostCallback();
+		this.observers.forEach(({mutation}) => mutation.disconnect());
+	}
+}
+
+function hasValidMutation(record: MutationRecord, mutation: Mutation) {
+	const {addedNodes, removedNodes, target} = record;
+	const {attributes, nodeName} = mutation;
+
+	// Is a removal or added mutation with type childList
+
+	if (removedNodes.length > 0 || addedNodes.length > 0) {
+		const [node] = removedNodes.length > 0 ? removedNodes : addedNodes;
```

Comment:
```suggestion
      const hasRemovedNodes = removedNodes.length > 0;
   
	if (hasRemovedNodes || addedNodes.length > 0) {
		const [node] = hasRemovedNodes ? removedNodes : addedNodes;
```

Reply:
This is more of just a small nit and probably a personal preference, since we check it multiple times in a row I typically store it as a variable. If we later decide we want to check if its length is greater than 1, then we update 2 places rather than one. This is an extremely small nit-pick now that I think about it 😂

Reply:
😂 oh ok, I think now I understand, well I think to keep it that way for now 😅